### ISSUE TITLE
kvars should be functions

### ISSUE THREAD

**Issue by gridaphobe** [ucsd-progsys/liquid-fixpoint#160]
I find it quite disturbing that the following is UNSAFE

```
// This qualifier saves the day; solve constraints WITHOUT IT
qualif Zog(v:a) : (10 <= v)

bind 0 x : {v1 : int | v1 = 10}
bind 1 y : {v2 : int | v2 = 20}
bind 2 a : {v3 : int | $k0    }
      
constraint:
  env [0]
  lhs {v4 : int | v4 = x}
  rhs {v5 : int | $k0   }
  id 1 tag []

constraint:
  env [1]
  lhs {v6 : int | v6 = y}
  rhs {v7 : int | $k0   }
  id 2 tag []

constraint:
  env [2]
  lhs {v8 : int | v8 = a  }
  rhs {v9 : int | 10 <= v9}
  id 3 tag []

wf:
  env [ ]
  reft {v10 : int | $k0}
```

but if I replace all the `vN`s with `v`, it becomes SAFE.

The two programs should be alpha-equivalent but they clearly aren't. I guess this is because we solve `$k0` for a concrete predicate, so it has to refer to a **specific** `vN`. Instead, we should (conceptually) solve it for a function, i.e. we should think of

```
{v:int | k}`
```

as

```
{v:int | k v}
```

I think this is exactly the same issue @ranjitjhala was talking about earlier today.

**Comment by nikivazou** [comment]
Can we do that by adding one more pending substitution at the well
formedness constraint when we create the k variables?

**Comment by gridaphobe** [comment]
Yes, I think that would work. But @ranjitjhala's suggestion yesterday was to do away with the pending substitutions altogether, which sounds much cleaner to me.

**Comment by nikivazou** [comment]
Sorry, I didn't follow the yesterday discussion, but doing away with
pending substitutions sounds like pushing abstract refinements at fixpoint.
If this makes any sense.

**Comment by gridaphobe** [comment]
Yes, that sounds about right.

**Comment by ranjitjhala** [comment]
The only reason your thing seems disturbing is that you've gotten the constraints wrong (as @nikivazou points out) really, the RHS _should_ have substitutions `[v10 := v7]` and `[v10 := v9]` respectively, and otherwise the constraints are simply malformed (and should have been rejected as such.)

(Thankfully, with `--native` they _are_ rejected -- or rather, it crashes right now, as I'd expect it to.)

---

At any rate, replacing the _pending substitutions_ with _functions_ is probably going to make this more obvious, but @bmcfluff -- we will have to figure out / re-formalize (and then re-implement) how constraint generation works in LH and Rsc and elsewhere, which will be a non-trivial thing. That said, the place to begin is to start with the PLDI 2008 paper and rewrite the constraint generation to use the "function" style instead of the substitutions.
